slack/web/classes/dialogs.py

```diff
@@ -0,0 +1,453 @@
+            self.options = None
+        self.optional = optional
+        self.value = value
+        self.placeholder = placeholder
+
+    @JsonValidator("options attribute cannot exceed 100 items")
+    def options_length(self):
+        return self.data_source != "static" or len(self.options) < 100
+
+    def get_json(self) -> dict:
+        json = super().get_json()
+        if self.data_source == "static":
+            json[self.property_key] = extract_json(self.options, OptionTypes, "label")
+        return json
+
+
+class StaticDialogDropdown(DialogDropdown):
+    data_source = "static"
+    property_key = "options"
+
+    def __init__(self, *, name: str, label: str, options: Iterable[Option], **kwargs):
+        super().__init__(name=name, label=label, options=options, **kwargs)
+
+
+class GroupedDialogDropdown(DialogDropdown):
+    data_source = "static"
+    property_key = "options_group"
+
+    def __init__(
+        self, *, name: str, label: str, options: Iterable[OptionGroup], **kwargs
+    ):
+        super().__init__(name=name, label=label, options=options, **kwargs)
+
+
+class DynamicDialogDropdown(DialogDropdown):
+    property_key = "options"
+
+    def __init__(self, *, name: str, label: str, source: str, **kwargs):
+        self.source = source
+        super().__init__(name=name, label=label, **kwargs)
+
+    @property
+    def data_source(self) -> str:
+        return self.source
+
+    @EnumValidator("source", DynamicDropdownTypes)
+    def source__valid(self):
+        return self.source in DynamicDropdownTypes
+
+
+class ExternalDialogDropdown(DialogDropdown):
+    data_source = "external"
+    property_key = "options"
+
+    @property
+    def attributes(self) -> Set[str]:
+        return super().attributes.union({"min_query_length"})
+
+    def __init__(
+        self,
+        *,
+        name: str,
+        label: str,
+        selected_option: Option = None,
+        min_query_length: int = None,
+        **kwargs,
+    ):
+        super().__init__(name=name, label=label, **kwargs)
+        self.selected_option = selected_option
+        self.min_query_length = min_query_length
+
+    def get_json(self) -> dict:
+        json = super().get_json()
+        if self.selected_option is not None:
+            # selected_option JSON must be a list, even though it will only be one item
+            json["selected_option"] = extract_json(
+                [self.selected_option], Option, "label"
+            )
+        return json
+
+
+class DialogBuilder(JsonObject):
+    _callback_id: Optional[str]
+    _elements: List[Union[DialogTextComponent, DialogDropdown]]
+    _submit_label: Optional[str]
+    _notify_on_cancel: bool
+    _state: Optional[str]
+
+    def __init__(self):
+        self._title = None
+        self._callback_id = None
+        self._elements = []
+        self._submit_label = None
+        self._notify_on_cancel = False
+        self._state = None
+
+    def title(self, title: str) -> "DialogBuilder":
```

Comment:
This annotation could be skipped, _or_ could `from future import annotations` in order to drop the string casing, although that does have performance implications on 3.6 I believe